### PR TITLE
Logging and metrics changes for distributed training

### PR DESCRIPTION
**Pull request by scarecrow1123**
As discussed at #2536 , this is the first PR to bring in `DistributedDataParallel` support. To start with this contains refactored implementations of `prepare_global_logging` and `training_util.get_metrics` methods.

* Both changes involve adding `rank` and `world_size` as optional attributes.

* Wrt logging, in the distributed case, each worker will log to their respective files with names `stdout_{rank}.log`/`stderr_{rank}.log`. 

* Logs to stdout/stderr are only from the master(rank=0) process. `tqdm` gets updated with metrics aggregated across all workers

* With these changes to logging, `TeeLogger` becomes unnecessary IMO and hence I took the liberty to remove it.

* `get_metrics` does a `dist.all_reduce` to aggregate all the provided metrics

P.S: pytest failures are yet to be fixed

### CODE REVIEW

File: allennlp/common/tee_logger.py [hunk 1]
```diff
@@ -1,63 +0,0 @@
...
-        sys.stderr = TeeLogger("stdout.log", sys.stderr)
-    """
-
-    def __init__(
-        self, filename: str, terminal: TextIO, file_friendly_terminal_output: bool
-    ) -> None:
-        self.terminal = terminal
-        self.file_friendly_terminal_output = file_friendly_terminal_output
-        parent_directory = os.path.dirname(filename)
-        os.makedirs(parent_directory, exist_ok=True)
```
Comment (DeNeutoy) [inline allennlp/common/tee_logger.py]: This call is missing in your replacement function, which is why all the tests are failing. Just add this and then I think that should kill most of the CI output :)

File: allennlp/training/util.py [hunk 2]
```diff
@@ -378,17 +379,31 @@ def rescale_gradients(model: Model, grad_norm: Optional[float] = None) -> Option
...
     """
     Gets the metrics but sets ``"loss"`` to
     the total loss divided by the ``num_batches`` so that
     the ``"loss"`` metric is "average loss per batch".
     """
     metrics = model.get_metrics(reset=reset)
     metrics["loss"] = float(total_loss / num_batches) if num_batches > 0 else 0.0
-    return metrics
+
+    if world_size > 1:
```
Comment (DeNeutoy) [inline allennlp/training/util.py]: I think this needs to be before the bit above where we replace the `metrics["loss"]` with a float, as it won't have been all_reduced (and then subsequently won't be all reduced correctly, as we've lost the reference to the tensor)

**Comment by DeNeutoy** [comment]
Also, as you're going to be submitting several PRs, here are a couple of development tips for allennlp which I find useful:

- You can run the CI checks locally using the scripts in the `scripts` directory, e.g `bash scripts/black.sh`, `bash scripts/flake8.sh`, `bash scripts/flake8.sh`. 
- Typically I run only tests I think I might have broken locally, e.g `pytest allennlp/tests/the/one/I/broke` otherwise it takes _super_ long.
- In our CI we have checks for the allennlp docs. In general don't worry about these, as we will always be able to tell you how to fix it, but basically we check that we have docs for all public modules, and that we don't have docs for modules which don't exist. For example, this PR is failing the docs check because you deleted the `TeeLogger`, but the docs are still present here: 

https://github.com/allenai/allennlp/blob/master/doc/api/allennlp.common.tee_logger.rst

Hopefully adding/removing them is self-explanatory, but getting them to build correctly can be a pain (sphinx is very fiddly). If you ever have trouble with that, let us know, because there's lots of sphinx rules i've internalised inside my head haha.

**Comment by brendan-ai2** [comment]
One clarification, if we literally `git revert 7afe971` that revert will be carried into `master` when we later merge all of this work. So I suspect @DeNeutoy was really suggesting something like:

1. Make a new branch `fix` based off of `torch-distributed`.
2. `git cherry-pick` in just the desired commits from `scarecrow1123:ddp-1`.
3. Then check out `scarecrow1123:ddp-1` and `git reset --hard fix`.

That will ensure `scarecrow1123:ddp-1` contains just the desired commits without the harmful revert. I suspect there's a better way to do step 2 than cherry picking in the commits individually, but this is what I came up with off the top of my head.

File: allennlp/common/tee_logger.py (hunk 1 above)
Comment (scarecrow1123) [reply allennlp/common/tee_logger.py]: The actual issue was with calling `logging.shutdown`. According to the [docs](https://docs.python.org/3/library/logging.html?highlight=logging%20shutdown#logging.shutdown) the log handlers are properly flushed/closed by default at module exit. Hence it is not needed and the tests pass now. 

`TeeLogger's` API received a filepath and hence it was necessary to create necessary directories(`TeeLogger` doesn't assume `parent_directory` to be the `serialization_dir`) . But in the current change, everything is taken care in `prepare_global_logging` which assumes the directory is always `serialization_dir` and hence this need not be done.

**Comment by scarecrow1123** [comment]
> it will make rebasing/merging against master easier if we don't pull in diffs from there as we go

> That will ensure scarecrow1123:ddp-1 contains just the desired commits without the harmful revert. I suspect there's a better way to do step 2 than cherry picking in the commits individually, but this is what I came up with off the top of my head.

Apologies for the goof up. I've modified the tree to be clean thanks to @brendan-ai2 's suggestions.

File: allennlp/training/util.py (hunk 2 above)
Comment (scarecrow1123) [reply allennlp/training/util.py]: Sorry, I don't quite get your point here. `total_loss` is a float and I'm not sure what reference we'd be losing. Essentially, `total_loss / num_batches` give the effective loss in a worker and we need to log the average loss across workers. Aggregation happens for all the metrics in the `dict` and hence the effective loss is being added there before the `for` loop. 
Could you elaborate a bit here?

Comment (DeNeutoy) [reply allennlp/training/util.py]: Ah I see - the loss has already been all-reduced for the backward pass so here we only care about the value.

Comment (scarecrow1123) [reply allennlp/training/util.py]: Strictly speaking, it is the gradients that have been all-reduced during the backward pass.

**Comment by scarecrow1123** [comment]
@DeNeutoy Let me know if there is anything else to be changed here and if this is good to go into the branch. The changes for the next PR is underway.

**Pull request merged**